pkg/embeddedcluster/monitor.go

```diff
@@ -0,0 +1,139 @@
+package embeddedcluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/replicatedhq/embedded-cluster-operator/api/v1beta1"
+	apptypes "github.com/replicatedhq/kots/pkg/app/types"
+	statetypes "github.com/replicatedhq/kots/pkg/appstate/types"
+	"github.com/replicatedhq/kots/pkg/k8sutil"
+	"github.com/replicatedhq/kots/pkg/logger"
+	"github.com/replicatedhq/kots/pkg/store"
+	storetypes "github.com/replicatedhq/kots/pkg/store/types"
+	"k8s.io/client-go/kubernetes"
+)
+
+// StartInstallationMonitor starts a goroutine that monitors the embedded cluster installation
+// and starts the upgrade process if necessary.
+func StartInstallationMonitor(ctx context.Context) error {
+	clientset, err := k8sutil.GetClientset()
+	if err != nil {
+		return fmt.Errorf("failed to get kubeclient: %w", err)
+	}
+	if isembedded, err := IsEmbeddedCluster(clientset); err != nil {
+		return fmt.Errorf("failed to check if embedded: %w", err)
+	} else if !isembedded {
+		return nil
+	}
+	mon := monitor{
+		store:     store.GetStore(),
+		clientset: clientset,
+	}
+	go mon.start(ctx)
+	return nil
+}
+
+// monitor is a struct that groups all methods needed to monitor the embedded cluster installation.
+type monitor struct {
+	store     store.Store
+	clientset *kubernetes.Clientset
+}
+
+// getApp returns the app deployed on top of the embedded cluster.
+func (m *monitor) getApp(ctx context.Context) (*apptypes.App, error) {
+	apps, err := m.store.ListInstalledApps()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list installed apps: %w", err)
+	} else if len(apps) == 0 {
+		return nil, nil
+	}
+	return apps[0], nil
+}
+
+// maybeStartUpgrade checks if the embedded cluster is in a state that requires an upgrade. If so,
+// it starts the upgrade process. We only start an upgrade if the following conditions are met:
+// - We have an app deployed on top of the embedded cluster.
+// - The deployed app version is in ready state.
+// - The app has an embedded cluster configuration.
+// - The app embedded cluster configuration differs from the current embedded cluster config.
+func (m *monitor) maybeStartUpgrade(ctx context.Context) error {
+	app, err := m.getApp(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get app: %w", err)
+	} else if app == nil {
+		return nil
+	}
+	cid, err := m.store.GetClusterIDFromSlug("this-cluster")
+	if err != nil {
+		return fmt.Errorf("failed to get cluster id: %w", err)
+	}
+	version, err := m.store.GetCurrentDownstreamVersion(app.ID, cid)
+	if err != nil {
+		return fmt.Errorf("failed to get downstream version: %w", err)
+	}
+	appv, err := m.store.GetAppVersion(app.ID, version.Sequence)
+	if err != nil {
+		return fmt.Errorf("failed to get app version: %w", err)
+	}
+	kinds := appv.KOTSKinds
+	notDeployed := version.Status != storetypes.VersionDeployed
+	noClusterConfig := kinds == nil || kinds.EmbeddedClusterConfig == nil
+	if notDeployed || noClusterConfig {
+		return nil
+	}
+	status, err := m.store.GetAppStatus(app.ID)
```

Comment:
i would move this up a bit to save a few unnecessary DB hits when the app isn't ready yet.

Reply:
we no longer check the app status here